File: apps/cptr/cptr/services/mcp_traffic.py
"""Bounded in-memory MCP traffic telemetry state.

The traffic subsystem records only an allowlisted operational envelope. It is
intentionally non-durable and never stores request payloads, tool arguments,
results, headers, credentials, prompts, or arbitrary exception messages.
"""

from __future__ import annotations

import asyncio
import os
import time
from collections import deque
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class McpTrafficClient(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1, max_length=128)
    label: str = Field(min_length=1, max_length=80)
    version: str | None = Field(default=None, max_length=64)
    session_name: str | None = Field(default=None, max_length=160)
    model: str | None = Field(default=None, max_length=120)
    workspace_id: str | None = Field(default=None, max_length=200)
    workspace_name: str | None = Field(default=None, max_length=160)


class McpTrafficEvent(BaseModel):
    model_config = ConfigDict(extra="forbid")

    version: Literal[1]
    event_id: str = Field(min_length=8, max_length=128)
    sequence: int = Field(ge=1)
    event_type: Literal[
        "session_opened",
        "session_closed",
        "request_started",
        "request_finished",
        "request_failed",
        "tool_started",
        "tool_finished",
        "tool_failed",
    ]
    timestamp_ms: int = Field(ge=0)
    session_id: str | None = Field(default=None, max_length=128)
    client: McpTrafficClient
    request_id: str | None = Field(default=None, max_length=128)
    correlation_id: str | None = Field(default=None, max_length=128)
    method: str | None = Field(default=None, max_length=128)
    tool_name: str | None = Field(default=None, max_length=256)
    status: Literal["started", "complete", "error", "connected", "disconnected"]
    duration_ms: int | None = Field(default=None, ge=0, le=86_400_000)
    request_bytes: int | None = Field(default=None, ge=0, le=100_000_000)
    response_bytes: int | None = Field(default=None, ge=0, le=100_000_000)
    error_code: (
        Literal[
            "timeout",
            "validation_error",
            "unauthorized",
            "tool_error",
            "transport_error",
            "internal_error",
        ]
        | None
    ) = None


class McpTrafficBatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    events: list[McpTrafficEvent] = Field(min_length=1, max_length=100)


def _bounded_env_int(name: str, default: int, minimum: int, maximum: int) -> int:
    raw = os.getenv(name)
    if raw is None:
        return default
    try:
        value = int(raw)
    except (TypeError, ValueError):
        return default
    return max(minimum, min(value, maximum))


class McpTrafficStore:
    """Bounded mutable topology state protected by one asyncio lock."""

    def __init__(
        self,
        *,
        max_events: int = 500,
        max_sessions: int = 128,
        subscriber_queue_size: int = 128,
        session_ttl_seconds: int = 300,
    ) -> None:
        self.max_events = max(1, int(max_events))
        self.max_sessions = max(1, int(max_sessions))
        self.subscriber_queue_size = max(1, int(subscriber_queue_size))
        self.session_ttl_ms = max(1, int(session_ttl_seconds)) * 1000
        self.max_active_requests = self.max_events
        self.max_clients = max(8, min(self.max_events, self.max_sessions * 2))
        self._dedupe_capacity = max(32, self.max_events * 4)

        self._lock = asyncio.Lock()
        self._events: deque[dict[str, object]] = deque(maxlen=self.max_events)
        self._seen_event_ids: deque[str] = deque()
        self._seen_event_id_set: set[str] = set()
        self._sessions: dict[str, dict[str, object]] = {}
        self._clients: dict[str, dict[str, object]] = {}
        self._active_requests: dict[str, dict[str, object]] = {}
        self._subscribers: set[asyncio.Queue[dict[str, object]]] = set()
        self._ingestion_sequence = 0
        self._slow_subscriber_drops = 0
        self._session_evictions = 0
        self._request_evictions = 0
        self._expired_sessions = 0

    def subscribe(self) -> asyncio.Queue[dict[str, object]]:
        queue: asyncio.Queue[dict[str, object]] = asyncio.Queue(maxsize=self.subscriber_queue_size)
        self._subscribers.add(queue)
        return queue

    def unsubscribe(self, queue: asyncio.Queue[dict[str, object]]) -> None:
        self._subscribers.discard(queue)

    async def ingest(self, events: list[McpTrafficEvent]) -> dict[str, int]:
        accepted = 0
        duplicates = 0
        dropped = 0
        async with self._lock:
            for event in events:
                if event.event_id in self._seen_event_id_set:
                    duplicates += 1
                    continue

                self._remember_event_id(event.event_id)
                self._ingestion_sequence += 1
                projected = event.model_dump(mode="json")
                projected["ingestion_sequence"] = self._ingestion_sequence
                self._events.append(projected)
                self._apply_event(event)
                self._fan_out(projected)
                accepted += 1

            self._prune_empty_placeholders()
            self._prune_clients()
        return {"accepted": accepted, "duplicates": duplicates, "dropped": dropped}

    async def snapshot(self) -> dict[str, object]:
        async with self._lock:
            clients = []
            for client_id in sorted(self._clients):
                client = self._clients[client_id]
                clients.append(
                    {
                        "id": client_id,
                        "label": client["label"],
                        "version": client["version"],
                        "session_name": client["session_name"],
                        "model": client["model"],
                        "workspace_id": client["workspace_id"],
                        "workspace_name": client["workspace_name"],
                        "active_sessions": self._active_session_count(client_id),
                        "active_requests": self._active_request_count(client_id),
                        "total_requests": client["total_requests"],
                        "errors": client["errors"],
                        "last_seen": client["last_seen"],
                        "last_tool": client["last_tool"],
                    }
                )

            sessions = [
                {
                    "session_id": session_id,
                    "client_id": session["client_id"],
                    "connected_at": session["connected_at"],
                    "last_seen": session["last_seen"],
                }
                for session_id, session in sorted(self._sessions.items())
            ]

            return {
                "version": 1,
                "sequence": self._ingestion_sequence,
                "center": {"id": "cptr-mcp", "label": "CPTR MCP", "status": "online"},
                "clients": clients,
                "sessions": sessions,
                "events": list(self._events),
                "stream_health": {
                    "subscriber_count": len(self._subscribers),
                    "slow_subscriber_drops": self._slow_subscriber_drops,
                    "session_evictions": self._session_evictions,
                    "request_evictions": self._request_evictions,
                    "expired_sessions": self._expired_sessions,
                    "event_capacity": self.max_events,
                    "session_capacity": self.max_sessions,
                },
            }

    async def expire_stale_sessions(self, now_ms: int | None = None) -> int:
        current = int(time.time() * 1000) if now_ms is None else int(now_ms)
        async with self._lock:
            stale = [
                session_id
                for session_id, session in self._sessions.items()
                if current - int(session["last_seen"]) > self.session_ttl_ms
            ]
            for session_id in stale:
                self._sessions.pop(session_id, None)
            self._expired_sessions += len(stale)
            self._prune_clients()
            return len(stale)

    def _remember_event_id(self, event_id: str) -> None:
        self._seen_event_ids.append(event_id)
        self._seen_event_id_set.add(event_id)
        while len(self._seen_event_ids) > self._dedupe_capacity:
            expired = self._seen_event_ids.popleft()
            self._seen_event_id_set.discard(expired)

    def _touch_client(self, event: McpTrafficEvent) -> dict[str, object]:
        client_id = event.client.id
        state = self._clients.get(client_id)
        if state is None:
            state = {
                "label": event.client.label,
                "version": event.client.version,
                "session_name": event.client.session_name,
                "model": event.client.model,
                "workspace_id": event.client.workspace_id,
                "workspace_name": event.client.workspace_name,
                "total_requests": 0,
                "errors": 0,
                "last_seen": event.timestamp_ms,
                "last_tool": None,
            }
            self._clients[client_id] = state
        else:
            state["label"] = event.client.label
            state["version"] = event.client.version
            state["session_name"] = event.client.session_name or state["session_name"]
            state["model"] = event.client.model or state["model"]
            state["workspace_id"] = event.client.workspace_id or state["workspace_id"]
            state["workspace_name"] = event.client.workspace_name or state["workspace_name"]
            state["last_seen"] = max(int(state["last_seen"]), event.timestamp_ms)
        if event.tool_name:
            state["last_tool"] = event.tool_name
        return state

    def _apply_event(self, event: McpTrafficEvent) -> None:
        client = self._touch_client(event)
        if event.session_id and event.session_id in self._sessions:
            self._sessions[event.session_id]["last_seen"] = max(
                int(self._sessions[event.session_id]["last_seen"]), event.timestamp_ms
            )
            self._sessions[event.session_id]["client_id"] = event.client.id

        if event.event_type == "session_opened" and event.session_id:
            if event.session_id not in self._sessions and len(self._sessions) >= self.max_sessions:
                oldest = min(
                    self._sessions.items(),
                    key=lambda item: (int(item[1]["last_seen"]), item[0]),
                )[0]
                self._sessions.pop(oldest, None)
                self._session_evictions += 1
            self._sessions[event.session_id] = {
                "client_id": event.client.id,
                "connected_at": event.timestamp_ms,
                "last_seen": event.timestamp_ms,
            }
        elif event.event_type == "session_closed" and event.session_id:
            self._sessions.pop(event.session_id, None)
        elif event.event_type == "request_started" and event.request_id:
            if (
                event.request_id not in self._active_requests
                and len(self._active_requests) >= self.max_active_requests
            ):
                oldest_request = min(
                    self._active_requests.items(),
                    key=lambda item: (int(item[1]["started_at"]), item[0]),
                )[0]
                self._active_requests.pop(oldest_request, None)
                self._request_evictions += 1
            self._active_requests[event.request_id] = {
                "client_id": event.client.id,
                "started_at": event.timestamp_ms,
                "tool_name": event.tool_name,
            }
        elif event.event_type in {"request_finished", "request_failed"}:
            if event.request_id:
                self._active_requests.pop(event.request_id, None)
            client["total_requests"] = int(client["total_requests"]) + 1
            if event.event_type == "request_failed":
                client["errors"] = int(client["errors"]) + 1

    def _fan_out(self, projected: dict[str, object]) -> None:
        for queue in tuple(self._subscribers):
            try:
                queue.put_nowait(projected)
                continue
            except asyncio.QueueFull:
                pass

            try:
                queue.get_nowait()
                self._slow_subscriber_drops += 1
            except asyncio.QueueEmpty:
                pass
            try:
                queue.put_nowait(projected)
            except asyncio.QueueFull:
                self._slow_subscriber_drops += 1

    def _active_session_count(self, client_id: str) -> int:
        return sum(1 for session in self._sessions.values() if session["client_id"] == client_id)

    def _active_request_count(self, client_id: str) -> int:
        return sum(
            1 for request in self._active_requests.values() if request["client_id"] == client_id
        )

    def _prune_empty_placeholders(self) -> None:
        removable = [
            client_id
            for client_id, state in self._clients.items()
            if int(state["total_requests"]) == 0
            and int(state["errors"]) == 0
            and self._active_session_count(client_id) == 0
            and self._active_request_count(client_id) == 0
        ]
        for client_id in removable:
            self._clients.pop(client_id, None)

    def _prune_clients(self) -> None:
        if len(self._clients) <= self.max_clients:
            return
        protected = {str(session["client_id"]) for session in self._sessions.values()} | {
            str(request["client_id"]) for request in self._active_requests.values()
        }
        removable = sorted(
            (
                (client_id, int(state["last_seen"]))
                for client_id, state in self._clients.items()
                if client_id not in protected
            ),
            key=lambda item: (item[1], item[0]),
        )
        for client_id, _ in removable:
            if len(self._clients) <= self.max_clients:
                break
            self._clients.pop(client_id, None)


mcp_traffic_store = McpTrafficStore(
    max_events=_bounded_env_int("CPTR_MCP_TRAFFIC_MAX_EVENTS", 500, 10, 10_000),
    max_sessions=_bounded_env_int("CPTR_MCP_TRAFFIC_MAX_SESSIONS", 128, 1, 2_000),
    subscriber_queue_size=_bounded_env_int("CPTR_MCP_TRAFFIC_SUBSCRIBER_QUEUE_SIZE", 128, 1, 2_000),
    session_ttl_seconds=_bounded_env_int("CPTR_MCP_TRAFFIC_SESSION_TTL_SECONDS", 300, 10, 86_400),
)

File: apps/cptr/cptr/frontend/src/lib/stores/mcp-traffic.ts
import type {
	McpTrafficClient,
	McpTrafficClientSnapshot,
	McpTrafficErrorCode,
	McpTrafficEvent,
	McpTrafficSessionSnapshot,
	McpTrafficSnapshot
} from '$lib/apis/mcp';

export type McpTrafficClientState = {
	id: string;
	label: string;
	version: string | null;
	sessionName: string | null;
	model: string | null;
	workspaceId: string | null;
	workspaceName: string | null;
	activeSessions: number;
	activeRequests: number;
	totalRequests: number;
	errors: number;
	lastSeen: number;
	lastTool: string | null;
};

export type McpTrafficSessionState = {
	sessionId: string;
	clientId: string;
	connectedAt: number;
	lastSeen: number;
};

export type McpTrafficActiveRequest = {
	requestId: string;
	correlationId: string | null;
	clientId: string;
	startedAt: number;
	method: string | null;
	toolName: string | null;
	sessionId: string | null;
	requestBytes: number | null;
};

export type McpRecentRequestRow = {
	requestId: string;
	correlationId: string | null;
	clientId: string;
	clientLabel: string;
	clientVersion: string | null;
	clientSessionName: string | null;
	clientModel: string | null;
	clientWorkspaceName: string | null;
	method: string | null;
	toolName: string | null;
	status: 'active' | 'complete' | 'error';
	startedAt: number;
	completedAt: number | null;
	durationMs: number | null;
	requestBytes: number | null;
	responseBytes: number | null;
	errorCode: McpTrafficErrorCode | null;
	sessionId: string | null;
};

export type McpTrafficState = {
	sequence: number;
	eventCapacity: number;
	sessionCapacity: number;
	clientCapacity: number;
	events: McpTrafficEvent[];
	clients: Record<string, McpTrafficClientState>;
	sessions: Record<string, McpTrafficSessionState>;
	activeRequests: Record<string, McpTrafficActiveRequest>;
	recentRequests: McpRecentRequestRow[];
	seenEventIds: string[];
};

export type McpTopologyNode = McpTrafficClientState & {
	x: number;
	y: number;
	angle: number;
	connected: boolean;
	active: boolean;
};

export type McpRequestOutcomeTotals = {
	total: number;
	success: number;
	failed: number;
	active: number;
};

export type McpRequestTimelineBucket = {
	startMs: number;
	endMs: number;
	success: number;
	failed: number;
	total: number;
};

function clientFromSnapshot(client: McpTrafficClientSnapshot): McpTrafficClientState {
	return {
		id: client.id,
		label: client.label,
		version: client.version,
		sessionName: client.session_name,
		model: client.model,
		workspaceId: client.workspace_id,
		workspaceName: client.workspace_name,
		activeSessions: client.active_sessions,
		activeRequests: client.active_requests,
		totalRequests: client.total_requests,
		errors: client.errors,
		lastSeen: client.last_seen,
		lastTool: client.last_tool
	};
}

function sessionFromSnapshot(session: McpTrafficSessionSnapshot): McpTrafficSessionState {
	return {
		sessionId: session.session_id,
		clientId: session.client_id,
		connectedAt: session.connected_at,
		lastSeen: session.last_seen
	};
}

function defaultClient(client: McpTrafficClient, timestamp: number): McpTrafficClientState {
	return {
		id: client.id,
		label: client.label,
		version: client.version,
		sessionName: client.session_name,
		model: client.model,
		workspaceId: client.workspace_id,
		workspaceName: client.workspace_name,
		activeSessions: 0,
		activeRequests: 0,
		totalRequests: 0,
		errors: 0,
		lastSeen: timestamp,
		lastTool: null
	};
}

function eventLimit(snapshot: McpTrafficSnapshot): number {
	return Math.max(1, snapshot.stream_health.event_capacity || snapshot.events.length || 1);
}

function boundedTail<T>(items: T[], limit: number): T[] {
	return items.length <= limit ? items : items.slice(items.length - limit);
}

function oldestKey<T extends { lastSeen?: number; startedAt?: number }>(
	record: Record<string, T>
): string | null {
	return (
		Object.entries(record).sort(
			([leftId, left], [rightId, right]) =>
				(left.lastSeen ?? left.startedAt ?? 0) - (right.lastSeen ?? right.startedAt ?? 0) ||
				leftId.localeCompare(rightId)
		)[0]?.[0] ?? null
	);
}

function pruneClients(clients: Record<string, McpTrafficClientState>, capacity: number): void {
	while (Object.keys(clients).length > capacity) {
		const removable = Object.values(clients)
			.filter((client) => client.activeSessions === 0 && client.activeRequests === 0)
			.sort((a, b) => a.lastSeen - b.lastSeen || a.id.localeCompare(b.id))[0];
		if (!removable) break;
		delete clients[removable.id];
	}
}

function upsertClient(
	clients: Record<string, McpTrafficClientState>,
	event: McpTrafficEvent
): McpTrafficClientState {
	const current = clients[event.client.id] ?? defaultClient(event.client, event.timestamp_ms);
	const next: McpTrafficClientState = {
		...current,
		label: event.client.label,
		version: event.client.version,
		sessionName: event.client.session_name ?? current.sessionName,
		model: event.client.model ?? current.model,
		workspaceId: event.client.workspace_id ?? current.workspaceId,
		workspaceName: event.client.workspace_name ?? current.workspaceName,
		lastSeen: Math.max(current.lastSeen, event.timestamp_ms),
		lastTool: event.tool_name ?? current.lastTool
	};
	clients[event.client.id] = next;
	return next;
}

function activeRequestFrom(event: McpTrafficEvent): McpTrafficActiveRequest | null {
	if (!event.request_id) return null;
	return {
		requestId: event.request_id,
		correlationId: event.correlation_id,
		clientId: event.client.id,
		startedAt: event.timestamp_ms,
		method: event.method,
		toolName: event.tool_name,
		sessionId: event.session_id,
		requestBytes: event.request_bytes
	};
}

function activeRow(
	request: McpTrafficActiveRequest,
	client: McpTrafficClientState
): McpRecentRequestRow {
	return {
		requestId: request.requestId,
		correlationId: request.correlationId,
		clientId: request.clientId,
		clientLabel: client.label,
		clientVersion: client.version,
		clientSessionName: client.sessionName,
		clientModel: client.model,
		clientWorkspaceName: client.workspaceName,
		method: request.method,
		toolName: request.toolName,
		status: 'active',
		startedAt: request.startedAt,
		completedAt: null,
		durationMs: null,
		requestBytes: request.requestBytes,
		responseBytes: null,
		errorCode: null,
		sessionId: request.sessionId
	};
}

function terminalRow(
	event: McpTrafficEvent,
	client: McpTrafficClientState,
	active?: McpTrafficActiveRequest
): McpRecentRequestRow | null {
	if (!event.request_id) return null;
	const startedAt = active?.startedAt ?? Math.max(0, event.timestamp_ms - (event.duration_ms ?? 0));
	return {
		requestId: event.request_id,
		correlationId: event.correlation_id ?? active?.correlationId ?? null,
		clientId: event.client.id,
		clientLabel: client.label,
		clientVersion: client.version,
		clientSessionName: client.sessionName,
		clientModel: client.model,
		clientWorkspaceName: client.workspaceName,
		method: event.method ?? active?.method ?? null,
		toolName: event.tool_name ?? active?.toolName ?? null,
		status: event.event_type === 'request_failed' ? 'error' : 'complete',
		startedAt,
		completedAt: event.timestamp_ms,
		durationMs: event.duration_ms,
		requestBytes: event.request_bytes ?? active?.requestBytes ?? null,
		responseBytes: event.response_bytes,
		errorCode: event.error_code,
		sessionId: event.session_id ?? active?.sessionId ?? null
	};
}

function replaceRecentRow(
	rows: McpRecentRequestRow[],
	row: McpRecentRequestRow,
	limit: number
): McpRecentRequestRow[] {
	return boundedTail([...rows.filter((item) => item.requestId !== row.requestId), row], limit);
}

function reconstructDerivedState(
	events: McpTrafficEvent[],
	clients: Record<string, McpTrafficClientState>,
	limit: number
): {
	activeRequests: Record<string, McpTrafficActiveRequest>;
	recentRequests: McpRecentRequestRow[];
} {
	const activeRequests: Record<string, McpTrafficActiveRequest> = {};
	let recentRequests: McpRecentRequestRow[] = [];

	for (const event of events) {
		const client = clients[event.client.id] ?? defaultClient(event.client, event.timestamp_ms);
		if (event.event_type === 'request_started') {
			const request = activeRequestFrom(event);
			if (!request) continue;
			activeRequests[request.requestId] = request;
			recentRequests = replaceRecentRow(recentRequests, activeRow(request, client), limit);
		} else if (event.event_type === 'request_finished' || event.event_type === 'request_failed') {
			const active = event.request_id ? activeRequests[event.request_id] : undefined;
			const row = terminalRow(event, client, active);
			if (event.request_id) delete activeRequests[event.request_id];
			if (row) recentRequests = replaceRecentRow(recentRequests, row, limit);
		}
	}

	return { activeRequests, recentRequests };
}

export function hydrateMcpTraffic(snapshot: McpTrafficSnapshot): McpTrafficState {
	const limit = eventLimit(snapshot);
	const events = boundedTail([...snapshot.events], limit);
	const clients = Object.fromEntries(
		snapshot.clients.map((client) => [client.id, clientFromSnapshot(client)])
	) as Record<string, McpTrafficClientState>;
	const sessions = Object.fromEntries(
		snapshot.sessions.map((session) => [session.session_id, sessionFromSnapshot(session)])
	) as Record<string, McpTrafficSessionState>;
	const derived = reconstructDerivedState(events, clients, limit);

	const sessionCapacity = Math.max(
		1,
		snapshot.stream_health.session_capacity || snapshot.sessions.length || 1
	);
	const clientCapacity = Math.max(8, limit + sessionCapacity);
	pruneClients(clients, clientCapacity);

	return {
		sequence: snapshot.sequence,
		eventCapacity: limit,
		sessionCapacity,
		clientCapacity,
		events,
		clients,
		sessions,
		activeRequests: derived.activeRequests,
		recentRequests: derived.recentRequests,
		seenEventIds: boundedTail(
			events.map((event) => event.event_id),
			limit * 2
		)
	};
}

export function applyMcpTrafficEvent(
	state: McpTrafficState,
	event: McpTrafficEvent
): McpTrafficState {
	if (event.ingestion_sequence <= state.sequence || state.seenEventIds.includes(event.event_id)) {
		return state;
	}

	const clients = Object.fromEntries(
		Object.entries(state.clients).map(([id, client]) => [id, { ...client }])
	) as Record<string, McpTrafficClientState>;
	const sessions = Object.fromEntries(
		Object.entries(state.sessions).map(([id, session]) => [id, { ...session }])
	) as Record<string, McpTrafficSessionState>;
	const activeRequests = Object.fromEntries(
		Object.entries(state.activeRequests).map(([id, request]) => [id, { ...request }])
	) as Record<string, McpTrafficActiveRequest>;
	let recentRequests = state.recentRequests.map((row) => ({ ...row }));
	const client = upsertClient(clients, event);

	if (event.session_id && sessions[event.session_id]) {
		sessions[event.session_id] = {
			...sessions[event.session_id],
			lastSeen: Math.max(sessions[event.session_id].lastSeen, event.timestamp_ms)
		};
	}

	switch (event.event_type) {
		case 'session_opened':
			if (event.session_id) {
				if (!sessions[event.session_id] && Object.keys(sessions).length >= state.sessionCapacity) {
					const evictedSessionId = oldestKey(sessions);
					if (evictedSessionId) {
						const evictedClient = clients[sessions[evictedSessionId].clientId];
						if (evictedClient) {
							evictedClient.activeSessions = Math.max(0, evictedClient.activeSessions - 1);
						}
						delete sessions[evictedSessionId];
					}
				}
				if (!sessions[event.session_id]) client.activeSessions += 1;
				sessions[event.session_id] = {
					sessionId: event.session_id,
					clientId: event.client.id,
					connectedAt: event.timestamp_ms,
					lastSeen: event.timestamp_ms
				};
			}
			break;
		case 'session_closed':
			if (event.session_id && sessions[event.session_id]) {
				const sessionClientId = sessions[event.session_id].clientId;
				delete sessions[event.session_id];
				const sessionClient = clients[sessionClientId];
				if (sessionClient)
					sessionClient.activeSessions = Math.max(0, sessionClient.activeSessions - 1);
			}
			break;
		case 'request_started': {
			const request = activeRequestFrom(event);
			if (request && !activeRequests[request.requestId]) {
				if (Object.keys(activeRequests).length >= state.eventCapacity) {
					const evictedRequestId = oldestKey(activeRequests);
					if (evictedRequestId) {
						const evictedClient = clients[activeRequests[evictedRequestId].clientId];
						if (evictedClient) {
							evictedClient.activeRequests = Math.max(0, evictedClient.activeRequests - 1);
						}
						delete activeRequests[evictedRequestId];
						recentRequests = recentRequests.filter((row) => row.requestId !== evictedRequestId);
					}
				}
				activeRequests[request.requestId] = request;
				client.activeRequests += 1;
				recentRequests = replaceRecentRow(
					recentRequests,
					activeRow(request, client),
					state.eventCapacity
				);
			}
			break;
		}
		case 'request_finished':
		case 'request_failed': {
			const active = event.request_id ? activeRequests[event.request_id] : undefined;
			if (event.request_id && active) delete activeRequests[event.request_id];
			client.activeRequests = Math.max(0, client.activeRequests - 1);
			client.totalRequests += 1;
			if (event.event_type === 'request_failed') client.errors += 1;
			const row = terminalRow(event, client, active);
			if (row) recentRequests = replaceRecentRow(recentRequests, row, state.eventCapacity);
			break;
		}
		case 'tool_started':
		case 'tool_finished':
		case 'tool_failed':
			if (event.tool_name) client.lastTool = event.tool_name;
			break;
	}

	pruneClients(clients, state.clientCapacity);

	return {
		...state,
		sequence: event.ingestion_sequence,
		events: boundedTail([...state.events, event], state.eventCapacity),
		clients,
		sessions,
		activeRequests,
		recentRequests,
		seenEventIds: boundedTail([...state.seenEventIds, event.event_id], state.eventCapacity * 2)
	};
}

export function requestOutcomeTotals(state: McpTrafficState): McpRequestOutcomeTotals {
	const total = Object.values(state.clients).reduce((sum, client) => sum + client.totalRequests, 0);
	const failed = Object.values(state.clients).reduce((sum, client) => sum + client.errors, 0);
	return {
		total,
		success: Math.max(0, total - failed),
		failed,
		active: Object.keys(state.activeRequests).length
	};
}

export function requestTimeline(
	state: McpTrafficState,
	nowMs: number,
	options: { windowMs?: number; bucketMs?: number } = {}
): McpRequestTimelineBucket[] {
	const requestedWindow = Number.isFinite(options.windowMs)
		? Math.floor(options.windowMs ?? 60_000)
		: 60_000;
	const requestedBucket = Number.isFinite(options.bucketMs)
		? Math.floor(options.bucketMs ?? 5_000)
		: 5_000;
	const windowMs = Math.min(3_600_000, Math.max(1_000, requestedWindow));
	const bucketMs = Math.min(windowMs, Math.max(1_000, requestedBucket));
	const bucketCount = Math.min(120, Math.max(1, Math.ceil(windowMs / bucketMs)));
	const effectiveWindow = bucketCount * bucketMs;
	const endMs = Math.max(0, Math.floor(nowMs));
	const startMs = endMs - effectiveWindow;
	const buckets: McpRequestTimelineBucket[] = Array.from({ length: bucketCount }, (_, index) => ({
		startMs: startMs + index * bucketMs,
		endMs: startMs + (index + 1) * bucketMs,
		success: 0,
		failed: 0,
		total: 0
	}));

	for (const event of state.events) {
		if (event.event_type !== 'request_finished' && event.event_type !== 'request_failed') continue;
		if (event.timestamp_ms < startMs || event.timestamp_ms > endMs) continue;
		const rawIndex = Math.floor((event.timestamp_ms - startMs) / bucketMs);
		const index = Math.min(bucketCount - 1, Math.max(0, rawIndex));
		const bucket = buckets[index];
		bucket.total += 1;
		if (event.event_type === 'request_failed') bucket.failed += 1;
		else bucket.success += 1;
	}

	return buckets;
}

export function topologyNodes(state: McpTrafficState): McpTopologyNode[] {
	const clients = Object.values(state.clients).sort((a, b) => a.id.localeCompare(b.id));
	const count = clients.length;
	if (count === 0) return [];

	return clients.map((client, index) => {
		const angle = -Math.PI / 2 + (index * 2 * Math.PI) / count;
		return {
			...client,
			x: 0.5 + Math.cos(angle) * 0.38,
			y: 0.5 + Math.sin(angle) * 0.34,
			angle,
			connected: client.activeSessions > 0,
			active: client.activeRequests > 0
		};
	});
}

export function recentRequestRows(state: McpTrafficState): McpRecentRequestRow[] {
	return [...state.recentRequests]
		.sort((a, b) => (b.completedAt ?? b.startedAt) - (a.completedAt ?? a.startedAt))
		.slice(0, 10);
}
